datafusion/common/src/scalar.rs

```diff
@@ -512,19 +516,19 @@ impl std::hash::Hash for ScalarValue {
 pub fn get_dict_value<K: ArrowDictionaryKeyType>(
     array: &dyn Array,
     index: usize,
-) -> (&ArrayRef, Option<usize>) {
-    let dict_array = as_dictionary_array::<K>(array).unwrap();
-    (dict_array.values(), dict_array.key(index))
+) -> Result<(&ArrayRef, Option<usize>)> {
+    let dict_array = as_dictionary_array::<K>(array)?;
```

Comment:
This is a public function and `as_dictionary_array` can fail depending on what `array` value the user passes to the function. The [`downcast_value`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/lib.rs#L73) macro called by `as_dictionary_array` returns an internal error if the cast is impossible. 

Returning a `Result` here makes sense because this error is recoverable.